web/hooks/use-persistent-state.ts

```diff
@@ -0,0 +1,54 @@
+import { useState, useEffect } from 'react'
+
+export type PersistenceOptions = {
+  store?: Storage
+  prefix: string
+  name: string
+}
+
+export const getKey = (prefix: string, name: string) => `${prefix}-${name}`
+
+export const saveState = (key: string, val: unknown, store: Storage) => {
+  if (val === undefined) {
+    store.removeItem(key)
+  } else {
+    store.setItem(key, JSON.stringify(val))
+  }
+}
+
+export const loadState = (key: string, store: Storage) => {
+  const saved = store.getItem(key)
+  if (typeof saved === 'string') {
+    try {
+      return JSON.parse(saved) as unknown
+    } catch (e) {
+      console.error(e)
+    }
+  } else {
+    return undefined
+  }
+}
+
+export const usePersistentState = <T>(
+  defaultValue: T,
+  persist?: PersistenceOptions
+) => {
+  const store = persist?.store
+  const key = persist ? getKey(persist.prefix, persist.name) : null
+  let initialValue
+  if (key != null && store != null) {
+    const saved = loadState(key, store) as T
```

Comment:
If the initialValue is loaded immediately from storage, the first render will be different on the client vs what is server-side rendered. I think NextJS will complain and weird glitches would happen.

You might have to start with a default or loading value (maybe just `undefined`), and have a useEffect() that sets the saved value.

I guess I would try what you have first and see that it does cause problems before switching.

Reply:
I guess the cookie store would solve this problem

Reply:
Good point, not sure why I didn't see warnings? My implementation appeared to be working.